File: deck/base/code/native/swift/runtime/json.swift
import Foundation

enum json {
    static func parse(_ text: String) -> Any {
        guard let data = text.data(using: .utf8),
              let value = try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed]) else { return NSNull() }
        return value
    }
    static func stringify(_ value: Any) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: value, options: [.fragmentsAllowed]) else { return "" }
        return String(data: data, encoding: .utf8) ?? ""
    }
    static func getField(_ value: Any, _ key: String) -> Any { return (value as? [String: Any])?[key] ?? NSNull() }
    static func getItem(_ value: Any, _ index: Int) -> Any {
        guard let array = value as? [Any], index >= 0, index < array.count else { return NSNull() }
        return array[index]
    }
    static func asNumber(_ value: Any) -> Double { return (value as? NSNumber)?.doubleValue ?? 0 }
    static func asText(_ value: Any) -> String { return value as? String ?? "" }
    static func asBoolean(_ value: Any) -> Bool { return (value as? NSNumber)?.boolValue ?? false }
    static func isNull(_ value: Any) -> Bool { return value is NSNull }
    static func makeObject() -> Any { return [String: Any]() }
    static func setField(_ value: Any, _ key: String, _ field: Any) -> Any {
        var dict = (value as? [String: Any]) ?? [:]
        dict[key] = field
        return dict
    }
    static func makeArray() -> Any { return [Any]() }
    static func pushItem(_ value: Any, _ item: Any) -> Any {
        var items = (value as? [Any]) ?? []
        items.append(item)
        return items
    }
    static func fromText(_ value: String) -> Any { return value }
    static func fromNumber(_ value: Double) -> Any { return value }
    static func fromBoolean(_ value: Bool) -> Any { return value }
    static func makeNull() -> Any { return NSNull() }
}

File: deck/base/code/native/rust/runtime/json.rs
mod json {
    use serde_json::Value;
    pub fn parse(text: String) -> Value { serde_json::from_str(&text).unwrap_or(Value::Null) }
    pub fn stringify(value: Value) -> String { serde_json::to_string(&value).unwrap_or_default() }
    pub fn get_field(value: Value, key: String) -> Value { value.get(&key).cloned().unwrap_or(Value::Null) }
    pub fn get_item(value: Value, index: i64) -> Value { value.get(index as usize).cloned().unwrap_or(Value::Null) }
    pub fn as_number(value: Value) -> f64 { value.as_f64().unwrap_or(0.0) }
    pub fn as_text(value: Value) -> String { value.as_str().map(|s| s.to_string()).unwrap_or_default() }
    pub fn as_boolean(value: Value) -> bool { value.as_bool().unwrap_or(false) }
    pub fn is_null(value: Value) -> bool { value.is_null() }
    pub fn make_object() -> Value { Value::Object(serde_json::Map::new()) }
    pub fn set_field(mut value: Value, key: String, field: Value) -> Value {
        if let Value::Object(map) = &mut value { map.insert(key, field); }
        value
    }
    pub fn make_array() -> Value { Value::Array(Vec::new()) }
    pub fn push_item(mut value: Value, item: Value) -> Value {
        if let Value::Array(items) = &mut value { items.push(item); }
        value
    }
    pub fn from_text(value: String) -> Value { Value::String(value) }
    pub fn from_number(value: f64) -> Value { serde_json::json!(value) }
    pub fn from_boolean(value: bool) -> Value { Value::Bool(value) }
    pub fn make_null() -> Value { Value::Null }
}
